firedrake/interpolation.py

```diff
@@ -166,10 +169,26 @@ def make_interpolator(expr, V, subset, access):
     elif len(arguments) == 1:
         if isinstance(V, firedrake.Function):
             raise ValueError("Cannot interpolate an expression with an argument into a Function")
-
         argfs = arguments[0].function_space()
+        target_mesh = V.ufl_domain()
+        source_mesh = argfs.mesh()
+        argfs_map = argfs.cell_node_map()
+        if target_mesh is not source_mesh:
+            if not isinstance(target_mesh.topology, firedrake.mesh.VertexOnlyMeshTopology):
+                raise NotImplementedError("Can only interpolate onto a Vertex Only Mesh")
+            if target_mesh.geometric_dimension() != source_mesh.geometric_dimension():
+                raise ValueError("Cannot interpolate onto a mesh of a different geometric dimension")
+            if not hasattr(target_mesh, "_parent_mesh") or target_mesh._parent_mesh is not source_mesh:
+                raise ValueError("Can only interpolate across meshes where the source mesh is the parent of the target")
```

Comment:
I think 
```
if target_mesh._parent_mesh is not source_mesh:
    raise ...
```
is good enough for now. (It is stronger than just checking geometric dimensions). When we generalise this for other `trans_mesh` cases, we would need more changes anyway.

Reply:
I need the `hasattr(target_mesh, "_parent_mesh")` otherwise you'll get an `AttributeError` if you try to interpolate across meshes in any case other than from parent mesh to immersed `VertexOnlyMesh`.

Reply:
From the point of view of interpolation you need `target_mesh.geometric_dimension() == source_mesh.geometric_dimension()` in all cases so I think it's valuable that it's at least pointed out for now. If there is ever any way that one managed to make `target_mesh._parent_mesh` have a different geometric dimension this would catch it. The fact I can't think of how that would happen doesn't mean one shouldn't check methinks.